docs/extend/cli_plugins.md

```diff
@@ -0,0 +1,98 @@
+     pull requests against that repo. If you see this file in
+     another repository, consider it read-only there, as it will
+     periodically be overwritten by the definitive file. Pull
+     requests which include edits to this file in other repositories
+     will be rejected.
+-->
+
+# Docker CLI Plugin Spec
+
+The `docker` CLI supports adding additional top-level subcommands as
+additional out-of-process commands which can be installed
+independently. These plugins run on the client side and should not be
+confused with "plugins" which run on the server.
+
+This document contains information for authors of such plugins.
+
+## Requirements for CLI Plugins
+
+### Naming
+
+A valid CLI plugin name consists only of lower case letters `a-z`
+and the digits `0-9`. The leading character must be a letter. A valid
+name therefore would match the regex `^[a-z][a-z0-9]*$`.
+
+The binary implementing a plugin must be named `docker-$name` where
+`$name` is the name of the plugin. On Windows a `.exe` suffix is
+mandatory.
+
+## Required sub-commands
+
+A CLI plugin must support being invoked in at least these two ways:
+
+* `docker-$name docker-cli-plugin-metadata` -- outputs metadata about
+  the plugin.
+* `docker-$name [GLOBAL OPTIONS] $name [OPTIONS AND FURTHER SUB
+  COMMANDS]` -- the primary entry point to the plugin's functionality.
+
+A plugin may implement other subcommands but these will never be
+invoked by the current Docker CLI. However doing so is strongly
+discouraged: new subcommands may be added in the future without
+consideration for additional non-specified subcommands which may be
+used by plugins in the field.
+
+### The `docker-cli-plugin-metadata` subcommand
+
+When invoked in this manner the plugin must produce a JSON object
+(and nothing else) on its standard output and exit success (0).
+
+The JSON object has the following defined keys:
+* `SchemaVersion` (_string_) mandatory: must contain precisely "0.1.0".
+* `Vendor` (_string_) mandatory: contains the name of the plugin vendor/author. May be truncated to 11 characters in some display contexts.
+* `ShortDescription` (_string_) optional: a short description of the plugin, suitable for a single line help message.
+* `Version` (_string_) optional: the version of the plugin, this is considered to be an opaque string by the core and therefore has no restrictions on its syntax.
+* `URL` (_string_) optional: a pointer to the plugin's web page.
+
+A binary which does not correctly output the metadata
+(e.g. syntactically invalid, missing mandatory keys etc) is not
+considered a valid CLI plugin and will not be run.
+
+### The primary entry point subcommand
+
+This is the entry point for actually running the plugin. It maybe have
+options or further subcommands.
+
+#### Required global options
+
+A plugin is required to support all of the global options of the
+top-level CLI, i.e. those listed by `man docker 1` with the exception
+of `-v`.
```

Comment:
I fear cobra may be doing something tricky and short-circuiting through to the help command. I'll have a look.

Reply:
Quick test; looks like this is because we add the parent command (which is the actual command that's being called) this seems to work, but we'd have to check if that's the right thing to do;

```patch
diff --git a/cli-plugins/examples/helloworld/main.go b/cli-plugins/examples/helloworld/main.go
index cbe01593..95dd2f93 100644
--- a/cli-plugins/examples/helloworld/main.go
+++ b/cli-plugins/examples/helloworld/main.go
@@ -44,6 +44,7 @@ func main() {
                        Run: func(cmd *cobra.Command, args []string) {
                                fmt.Fprintf(dockerCli.Out(), "Hello %s!\n", who)
                        },
+                       Version: "0.0.2",
                }
                flags := cmd.Flags()
                flags.StringVar(&who, "who", "World", "Who are we addressing?")
diff --git a/cli-plugins/plugin/plugin.go b/cli-plugins/plugin/plugin.go
index c72a68bb..5745b2aa 100644
--- a/cli-plugins/plugin/plugin.go
+++ b/cli-plugins/plugin/plugin.go
@@ -91,6 +91,10 @@ func newPluginCommand(dockerCli *command.DockerCli, plugin *cobra.Command, meta
        }
        opts, flags := cli.SetupPluginRootCommand(cmd)
 
+       if plugin.Version != "" {
+               cmd.Version = plugin.Version
+       }
+       cmd.SetVersionTemplate(plugin.Name() + " version {{.Version}}\n")
        cmd.SetOutput(dockerCli.Out())
 
        cmd.AddCommand(
```

Reply:
Bigger issue (that I'm invesitgating now) is that `docker helloworld -v` gets short circuited by the cli so the plugin never gets an opportunity to run. Once I've sorted that I'll figure out what needs to be done in the plugin framework (you suggestions looks like a good start)